egui/src/grid.rs

```diff
@@ -182,6 +251,72 @@ impl GridLayout {
         cursor.min.x += frame_rect.width() + self.spacing.x;
     }
 
+    /// Paint a row.
+    pub(crate) fn paint_row(&self, min: Pos2, color: GuiColor, painter: &Painter) {
+        let color = color.pick(&self.style);
+        if let Some(height) = self.prev_state.row_height(self.row) {
+            // Paint background for coming row:
+            let size = Vec2::new(self.prev_state.full_width(self.spacing.x), height);
+            let rect = Rect::from_min_size(min, size);
+            let mut rect = rect.expand2(Vec2::new(2.0, 0.0));
+            rect.max += Vec2::new(1.0, 0.5 * self.spacing.y + 1.5);
+
+            painter.rect_filled(rect, 2.0, color);
+        }
+    }
+
+    /// Paint a column
+    pub(crate) fn paint_column(&self, col: usize, min: Pos2, color: GuiColor, painter: &Painter) {
+        let col_f = col as f32;
+
+        let color = color.pick(&self.style);
+
+        // Paint a column:
+        // Offset from the cursor to paint the col at the right spot.
+        let min_offset = min
+            + Vec2::new(
+                // Sum up all the previous widths and add the padding
+                self.prev_state.col_widths.iter().take(col).sum::<f32>()
+                    + (self.spacing.x + 1.0) * (col_f - 1.0),
+                -1.0,
+            );
+        let size = Vec2::new(
+            self.prev_col_width(col),
+            self.prev_state.full_height(self.spacing.y),
+        );
+        let size = size
+            + Vec2::new(
+                // Add padding
+                0.5 * self.spacing.x + 5.0,
+                3.0,
+            );
+
+        let rect = Rect::from_min_size(min_offset, size);
+
+        painter.rect_filled(rect, 2.0, color);
+    }
+
+    /// Paint the background for the next row & columns.
+    pub(crate) fn do_paint(&mut self, min: Pos2, painter: &Painter) {
+        for p in &self.color_spec.row_pickers {
+            if let Some(color) = p(self.row) {
+                self.paint_row(min, color, painter)
+            }
+        }
+        // Only paint columns when we're on the first row.
+        // They are painted for the entire grid.
+        if self.row == 0 {
+            for p in &self.color_spec.col_pickers {
+                // Iterate over columns
+                for col in 0..self.prev_state.dimensions().y {
+                    if let Some(color) = p(col) {
+                        self.paint_column(col, min, color, painter);
+                    }
+                }
```

Comment:
O(N²) !

Each column sums up the width of the previous columns, and you do that for each column.

Better to keep a running sum here, and then we don't need `struct Dimensions`:

Reply:
Fixed